### PR TITLE
docs: add Grid data provider examples

### PR DESCRIPTION
**Pull request by vursen**
Added a Grid example showing basically how to use data providers for lazy loading, sorting, and filtering items.

Fixes https://github.com/vaadin/docs/issues/79, https://github.com/vaadin/docs/issues/513

### CODE REVIEW

File: src/main/java/com/vaadin/demo/component/grid/GridDataProvider.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.Comparator;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+@Route("grid-data-provider")
+public class GridDataProvider extends Div {
+
+    // tag::snippet[]
+    private static class PersonFilter {
```
Comment (vursen) [inline src/main/java/com/vaadin/demo/component/grid/GridDataProvider.java:30]: I should probably consider moving these classes to separate files. The example swelled up quite a bit.

File: articles/components/grid/index.asciidoc [hunk 2]
```diff
@@ -454,6 +454,32 @@ include::{root}/src/main/java/com/vaadin/demo/component/grid/GridExternalFilteri
 ----
 --
 
+== Lazy Loading
+
+When you want to display a list of items that is too large to keep it whole in memory, or you want to load items from a database, data providers can be used to enable lazy loading.
```
Comment (vursen) [inline articles/components/grid/index.asciidoc:459]:
```suggestion
When you want to display a list of items that is too large to keep it whole in memory, or you want to load items from a database, data providers can be used to provide lazy loading.
```

Comment (vursen) [reply articles/components/grid/index.asciidoc:459]: Lazy Loading vs Pagination?

File: src/main/java/com/vaadin/demo/component/grid/PersonDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    protected Stream<Person> fetchFromBackEnd(Query<Person, PersonFilter> query) {
+        // A real app should use a real database or a service
+        // to fetch, filter and sort data.
+        Stream<Person> stream = DATABASE.stream();
+
+        // Filtering
+        if (query.getFilter().isPresent()) {
+            stream = stream.filter(filterPredicate(query.getFilter().get()));
```
Comment (bwajtr) [inline src/main/java/com/vaadin/demo/component/grid/PersonDataProvider.java:28]: The use of filterPredicate method feels too complicated here. For the sake of this example, I'd go simply with:
```
stream = stream.filter(person -> query.getFilter().get().test(person))
```

File: articles/components/grid/index.asciidoc (hunk 2 above)
Comment (bwajtr) [reply articles/components/grid/index.asciidoc:459]: I would stay with Lazy Loading. I recently had a customer which referred to this style of providing data to grids/comboboxes as "Lazy Loading" too....

Comment (bwajtr) [reply articles/components/grid/index.asciidoc:459]: Maybe the word "Pagination" could be added to the sentence below the header:

> When you want to display a list of items that is too large to keep it whole in memory, you want to implement pagination, or you want to load items from a database, data providers can be used to enable lazy loading.

**Review by bwajtr — APPROVED** [review]

**Pull request merged**